Src/IronPython/Runtime/PythonEncoding.cs

```diff
@@ -254,22 +286,104 @@ public class PythonEncoder : Encoder {
+                if (chars == null) throw new ArgumentNullException(nameof(chars));
+                if (bytes == null) throw new ArgumentNullException(nameof(bytes));
+                if (charIndex < 0) throw new ArgumentOutOfRangeException(nameof(charIndex));
+                if (charCount < 0) throw new ArgumentOutOfRangeException(nameof(charCount));
+                if (chars.Length - charIndex < charCount) throw new ArgumentOutOfRangeException(nameof(chars));
+                if (byteIndex < 0 || byteIndex > bytes.Length) throw new ArgumentOutOfRangeException(nameof(byteIndex));
+
+                var s = new string(chars, charIndex, charCount);
+                return GetBytes(s, bytes.AsSpan(byteIndex), flush);
+            }
+
+            // NLS workhorse
+            public override unsafe int GetBytes(char* chars, int charCount, byte* bytes, int byteCount, bool flush) {
+                if (chars == null) throw new ArgumentNullException(nameof(chars));
+                if (bytes == null) throw new ArgumentNullException(nameof(bytes));
+                if (charCount < 0) throw new ArgumentOutOfRangeException(nameof(charCount));
+                if (byteCount < 0) throw new ArgumentOutOfRangeException(nameof(byteCount));
+
+                var s = new string(chars, 0, charCount);
+                return GetBytes(s, new Span<byte>(bytes, byteCount), flush);
+            }
+
+            // used by IronPython
+            public int GetBytes(string s, int charIndex, int charCount, byte[] bytes, int byteIndex, bool flush) {
+                if (s == null) throw new ArgumentNullException(nameof(s));
+                if (bytes == null) throw new ArgumentNullException(nameof(bytes));
+                if (charIndex < 0) throw new ArgumentOutOfRangeException(nameof(charIndex));
+                if (charCount < 0) throw new ArgumentOutOfRangeException(nameof(charCount));
+                if (s.Length - charIndex < charCount) throw new ArgumentOutOfRangeException(nameof(s));
+                if (byteIndex < 0 || byteIndex > bytes.Length) throw new ArgumentOutOfRangeException(nameof(byteIndex));
+
+                if (charIndex != 0 || charCount != s.Length) {
+                    s = s.Substring(charIndex, charCount);
+                }
+                return GetBytes(s, bytes.AsSpan(byteIndex), flush);
+            }
+
+            private int GetBytes(string data, Span<byte> bytes, bool flush) {
                 var fbuf1 = GetPythonEncoderFallbackBuffer(_pass1encoder);
                 var fbuf2 = GetPythonEncoderFallbackBuffer(_pass2encoder);
-                fbuf1?.PrepareIncrement(chars, forEncoding: true);
+                fbuf1?.PrepareIncrement(data, forEncoding: true);
+
+                var chars = data.AsSpan();
 
-                int written = _pass1encoder.GetBytes(chars, charIndex, charCount, bytes, byteIndex, flush);
+                int written;
+                unsafe {
```

Comment:
.NET Core provides `Span` overloads for all(?) of these pointer based calls. I wonder if we could define extensions methods which would replicate the API surface for .NET Framework and then use the nicer `Span` syntax. This would just become:
```C#
int written = _pass1encoder.GetBytes(chars.Slice(0, chars.Length), bytes.Slice(0, bytes.Length), flush);
```
and we could add the following extension method:
```C#
#if !NETCOREAPP
internal static int GetBytes(this Encoder encoder, ReadOnlySpan<char> chars, Span<byte> bytes, bool flush) {
    unsafe {
        fixed (char* pChars = &MemoryMarshal.GetReference(chars))
        fixed (byte* pBytes = &MemoryMarshal.GetReference(bytes)) {
            return encoder.GetBytes(pChars, chars.Length, pBytes, bytes.Length, flush);
        }
    }
}
#endif
```

I think we could eliminate most (if not all) of the unsafe blocks this way.

Reply:
Great idea! Will get on it.

I think .NET 5 also has these. Will `NETCOREAPP` be defined for .NET 5?

Reply:
Yes, `NETCOREAPP` is also defined with .NET 5. I'm not sure if I'll be adding a .NET 5 target (unless there's some API surface that we want to use), however I'll probably try and get testing working on .NET 5 to make sure there are no issues with it.